### ISSUE TITLE
🔧 Make pyodbc an optional dependency (feature) through setuptools extras

### ISSUE THREAD

**Issue by cpcloud** [exasol/sqlalchemy-exasol#353]
# Summary

This is just a thought based on some light interaction with the library, but it would be nice if different dependencies for all the ways of connecting could be made optional somehow.

# Details

## Background & Context

@Nicoretti opened [a nice discussion about adding an ibis backend](https://github.com/ibis-project/ibis/discussions/6448), and it's likely we'll use this library, and ideally we can use the websockets API to avoid ODBC :)

However, it seems that the package unconditionally requires (and therefore installs) pyodbc, even if it never gets used.

Ideally we can avoid giving users a dependency they never use.

## Examples

I think the main changes would be that installation would look different:

```
$ pip install sqlalchemy-exasol[pyodbc]
$ pip install sqlalchemy-exasol[websockets]
$ pip install sqlalchemy-exasol # not sure what this would do, maybe an extra should be required?
```

## Related Issue(s)
* https://github.com/exasol/sqlalchemy-exasol/issues/402

**Comment by Nicoretti** [comment]
Hi @cpcloud,

Thanks for opening this issue.

## Summary 

>  ...ideally we can use the websockets API to avoid ODBC :)

**TL;DR: I think we are on the same page here ;)**

## Details 

Until recently the Exasol SQLA plugin was solely using ODBC, the websocket support was just added recently (Beta).
The long term plan, from what I know, is to make it the default. That said personally I would like to see it as the sole
solution when using SQLA. Still, if removing the ODBC based dialects is an option at some point is a more complex
question and needs further discussion and investigations.

### :spiral_notepad: Side Note
An SQLA based IBIS backend should be implemented using the Exasol websocket dialect/api.
If any issue with the websocket based dialect will surface we will address them.

**Comment by Nicoretti** [comment]
## Implementation Hints
 
[Poetry does not support extras/option based extension point registration](https://github.com/python-poetry/poetry/issues/1454). Therefore the optional registration of the dialects need to be addressed with an other mechanism, otherwise it may lead to unexpected behavior/errors. 


## Possible Solutions

### Provide different packages for each dialect

We can provide different packages for each dialect, so the main package still can use the `extras` mechanism,
due to the fact that the `extras` registration would be encapsulated within the dependencies itself.

### Check if the dialect feature was installed when the dialect is requested
We can add a runtime (import) based check if the required dialect `extra` is installed when the user
is requesting it.

**Comment by Nicoretti** [comment]
Discuss with @exa-eswar, to make `websocket` the default and add `odbc` as feature flag for version(s) >= `5.x.y`.
(Adding deprecation warnings in versions > `4.6.0`)

Until clarified we will use `4.6.0` for implementing the `IBIS` backend.
